pyop2/sparsity.pyx

```diff
@@ -56,304 +54,119 @@ cdef extern from "petsc.h":
-                    for i in range(cur_nrows):
-                        diag[c][row_offset + i].reserve(6*rarity)
-                        if alloc_diag and i < ncols:
-                            # Always allocate space for diagonal.
-                            # Note we only add the row_offset to the
-                            # index, not the inserted value, since
-                            # when we walk over the column maps we
-                            # don't add offsets.
-                            diag[c][row_offset + i].insert(i)
-                        if parallel:
-                            odiag[c][row_offset + i].reserve(6*rarity)
-                if extruded:
-                    add_entries_extruded(rset[r], rmap,
-                                         cset[c], cmap,
-                                         row_offset,
-                                         diag[c], odiag[c],
-                                         should_block)
-                else:
-                    add_entries(rset[r], rmap,
-                                cset[c], cmap,
-                                row_offset,
-                                diag[c], odiag[c],
-                                should_block)
-                restore_writeable(cmap, cflag)
-            # Increment only by owned rows
-            row_offset += rset[r].size * rdim
-            restore_writeable(rmap, rflag)
-
-    cdef np.ndarray[PetscInt, ndim=1] nnz = np.zeros(nrows, dtype=IntType)
-    cdef np.ndarray[PetscInt, ndim=1] onnz = np.zeros(nrows, dtype=IntType)
-    cdef np.ndarray[PetscInt, ndim=1] rowptr
-    cdef np.ndarray[PetscInt, ndim=1] colidx
-    cdef int nz, onz
-    if make_rowptr:
-        rowptr = np.empty(nrows + 1, dtype=IntType)
-        rowptr[0] = 0
+        # Sparsity is the block sparsity
+        nrows = rset.size
+        ncols = cset.size
+        preallocator.setLGMap(rmap=rset.scalar_lgmap, cmap=cset.scalar_lgmap)
+
+    preallocator.setSizes(size=((nrows, None), (ncols, None)),
+                          bsize=1)
+    preallocator.setUp()
+
+    if mixed:
+        for i, r in enumerate(rset):
+            for j, c in enumerate(cset):
+                maps = list(zip((m.split[i] for m in sparsity.rmaps),
+                                (m.split[j] for m in sparsity.cmaps)))
+                mat = preallocator.getLocalSubMatrix(isrow=rset.local_ises[i],
+                                                     iscol=cset.local_ises[j])
+                fill_with_zeros(mat, (r.cdim, c.cdim),
+                                maps,
+                                set_diag=((i == j) and sparsity._has_diagonal))
+                mat.assemble()
+                preallocator.restoreLocalSubMatrix(isrow=rset.local_ises[i],
+                                                   iscol=cset.local_ises[j],
+                                                   submat=mat)
+        preallocator.assemble()
```

Comment:
We use new PETSc technology to build preallocation, which greatly simplifies things.